bonsai-core/src/main/scala/com/stripe/bonsai/Layout.scala

```diff
@@ -4,31 +4,45 @@ import scala.language.implicitConversions
 
 import scala.reflect.ClassTag
 
+import java.io.{ DataOutput, DataInput }
+
 import com.stripe.bonsai.layout._
 
 trait Layout[A] {
   def newBuilder: VecBuilder[A]
 
+  def write(vec: Vec[A], out: DataOutput): Unit
+
+  def read(in: DataInput): Vec[A]
+
+  def isSafeToCast(vec: Vec[_]): Boolean
+
```

Comment:
is it worth having a `def rebuild(vec: Vec[_]): Vec[A]` method here? If we added that we could then also define `recast` here and remove some boilerplate from the 5 places we independently define recast below.